e2e: add test scenario to test with kOps

justinsb opened the pull request:
Uploading for discussion.

A simple e2e test that tests with a manifest that we provide. Solves the problem that we need to test with a configuration, treating that as part of our distribution (unless we want to test and support everything!).

Example issue: https://github.com/kubernetes/cloud-provider-gcp/pull/393

aojea commented on packages/manifests/direct/manifest.yaml:
```diff
@@ -0,0 +1,342 @@
+  - nodes
+  verbs:
+  - get
+  - update
```
you need patch too, for the nodeipam
```suggestion
  - update
  - patch
```

aojea replied:
This patches node.spec.PodCIDRs to CNI can use it as IPAM, Calico has its own IPAM, may be don't need it.
 Another possibility is that the KCM ipam kept working and patching the nodes.

justinsb replied:
I figured it out ... the nodes.patch permission is in the cloud-node-controller, but until #393 lands we aren't using the per-controller serviceaccounts (and therefore roles).  Of course we should also decide whether we want to use per-controller serviceaccounts, but that's sort of the bigger idea of this PR.  Even if we end up with two manifests - one for use-service-account-credentials=true and one for =false), they should be the manifests that the project wants to support.

I incorporated this, but I will also add a comment that it's only needed until we merge #393.

jprzychodzen commented on e2e/scenarios/kops-simple:
```diff
@@ -0,0 +1,145 @@
+echo "KOPS_FEATURE_FLAGS=${KOPS_FEATURE_FLAGS}"
+
+KUBETEST2_ARGS="kops --kops-binary-path=${BINDIR}/kops"
+KUBETEST2_ARGS="${KUBETEST2_ARGS} -v=2 --cloud-provider=${CLOUD_PROVIDER}"
```
How will this affect kubelet/kcm? For kubelet and KCM we want to set `--cloud-provider=external`.

justinsb replied:
So yes, I realize this is a little confusing.: These are the _kops_ arguments.  I obviously need a README here, but the idea is that we can test with kOps (and hopefully other production tools), and cloud-provider-gcp should control the manifest it is verified to run with (e.g. do we need nodes patch permission, as above).

The kops `--cloud-provider` flag did originally map to kubelet and kcm, but as of recent versions of k8s we will use cloud-provider=external there.  The flag is the user saying what cloud they want, and then kOps figures out the flags for the various components.

In kOps, I'd like to get us to a more modular approach where we don't have to create our own manifests for all the components, but rather cloud-provider-gcp can specify how it wants to be run, can easily test that manifest with kOps + other tools, and then kOps (and other tools) can run with that tested manifest.

cheftako replied:
I think a comment or README would be good. Also once we switch the disable cloud provider feature flag to beta, the KCM, KAS and Kubelet will all crash if you provide a gcp/gce cloud provider flag.

justinsb replied:
Added a comment!  And yes, kOps will map the flags to wherever they move to over time :-)

jprzychodzen commented on packages/manifests/direct/manifest.yaml:
```diff
@@ -0,0 +1,342 @@
+---
```
I guess this file could be regenerated by some simple script which cats appropriate manifests, right?

justinsb replied:
Yes - happy to do that ... I'm guessing https://github.com/kubernetes/cloud-provider-gcp/tree/master/deploy is the canonical source?  I wanted to keep this relatively self-contained at first, and one gotcha for unification is the Pod vs Daemonset for the c-c-m, but I think we can relatively easily create a script/tool that could generate both forms if that's the direction we agree.

My desired outcome is that kOps consumes a manifest that cloud-provider-gcp tests with.  Today that isn't the case, and I'd love to get these aligned for kOps.   I think it matters because we have functionality like #393 that isn't tested by default - I think that's fine if it's a configuration nobody uses, but I'm hoping we can make a stronger statement in cloud-provider-gcp about what configurations people should use.

It might be that statement is "thou shalt run as a static pod", that's more work for me as a consumer in the short term, but clarity and using a cross-tested configuration is more important in the long run.

jprzychodzen commented on e2e/scenarios/kops-simple:
```diff
@@ -0,0 +1,145 @@
+cd ${REPO_ROOT}
+export KO_DOCKER_REPO="${IMAGE_REPO}"
+ko build --tags ${IMAGE_TAG} --base-import-paths --push=true ./cmd/cloud-controller-manager/
+
```
I guess we want to build and push external credential provider - single binary which is used by kubelet as a shortliving process that generates GCR credentials.

This probably should be something like `ko build ./cmd/auth-provider-gcp`

justinsb replied:
Yes, good idea.  Do you know if we have any e2e tests (anywhere) that would fail if we did this "wrong"?   We do have a handful of e2e tests in the kOps repo itself, so the good news is that we don't have to get these upstream into k/k, we can easily `go test` some tests from this repo.

cheftako replied:
I know https://testgrid.k8s.io/google-gce#disable-cloud-provider&width=20 had a pretty broad set of failures before we got the pieces running correctly...

justinsb replied:
We are currently passing Conformance (skipping Serial) - I wonder if these tests would have to be built outside of k/k (because they are cloud provider specific)

justinsb commented:
> Could you explain how can we leverage this to remove /cluster directory?

So I think this can coexist with the cluster directory if desired, but if we wanted to get rid of the cluster directory....  We can have a few scenarios like this for the various OSS tools.  Those can run as periodics or against PRs, as desired.  kOps uses the kubetest2 framework, which is the kubernetes tooling which (hopefully) all the distros etc are using.

And indeed, https://github.com/kubernetes-sigs/kubetest2 shows that many OSS kubernetes tools are using kubetest2.  (Note that kubetest2-kops is in the kubernetes/kops repo because that was the advice at the time, I don't think anyone would mind moving it though, it's well isolated in its own module).  

kubetest2 doesn't immediately give us the ability to "just swap out" the tooling, but it does mean that all the scripts will look pretty similar.

So in particular, if we wanted to use this to get `cluster` out of this repo, we would use `kubetest2-gce` 
 (https://github.com/kubernetes-sigs/kubetest2/tree/master/kubetest2-gce) instead of using `kubetest-kops`.  Likely in a separate script, which is why I created a directory for "scenarios".

We would source the "gce" distro from wherever it is developed, just as we source kOps from kubernetes/kops in this PR.  Today, the "gce" distro is developed in k/k, but I could imagine us creating a repo like "kubernetes/kubeup-gce" and developing it out of tree.  I _think_ we could use the same approach we have here in k/k, where a script uses the upstream distro.

The tricky thing is that we want to replace the CCM manifest in kOps / kubeup-gce / etc with our own CCM manifest for the version under test.  That replacement mechanism is going to be specific to each "distro", but we could look at standardizing that also in kubetest2 (maybe?).  For reference, this PR depends on this WIP PR in kOps: https://github.com/kubernetes/kops/pull/14834 , which is a reasonable mechanism IMO but input is very welcome there on how explicit we should be (and I'm sure we'll debate it!)

justinsb commented:
I've updated the test to run with "vpc native" networking (kOps calls it "gce" networking).  I've incorporated @aojea 's suggestion for node.patch permission, and verified that we need this until #393 lands.

We had some debates in kOps about how to identify a "cluster addon" so that we could not install the built-in GCE CCM support when a user (or the test) was providing a GCE CCM.  In the end we settled on using labels, and we had an existing label so we used that.  That is the `addon.kops.k8s.io/name: gcp-cloud-controller.addons.k8s.io` label.  I'd like to see if we can get agreement on a label that doesn't include the kops.k8s.io name so that it can be more "neutral" - the intent is that these manifests can be used by all tooling.

I've also created a periodic job that should run these tests - that is https://github.com/kubernetes/test-infra/pull/28386.  These tests _add_ to the existing tests (and run every 24 hours, not on every PR).

My suggestion is that:

* We merge this and https://github.com/kubernetes/test-infra/pull/28386.  Hopefully we start to get some test signal at that stage.
* We can then undo the node.patch permission, verify that the tests fail, and then merge #393.

Then as additional follow-ons, I (or anyone else that wants to participate) can then:

* Work on unifying the manifest generation (as suggested by @jprzychodzen above)
* Work on adding a similar test/manifest for the external credential provider  (also as suggested by @jprzychodzen)
* Start figuring out whether we like the label approach and whether we can agree some labels project-wide
* Start figuring out what parameters we need to pass here from kOps to the CCM.  For example we pass the PodCIDR; we _might_ not need to, but assuming we do what is the right way to communicate these values?

justinsb commented:
I uploaded #450 which explores the first follow-up - how we should generate our preferred manifests for the configurations we intend to (primarily) support.

Also moved the manifest around in this PR to be the location where the files _will_ be if we do #450.

cheftako commented:
/lgtm
/approve

aojea commented on deploy/packages/default/manifest.yaml:
```diff
@@ -0,0 +1,356 @@
+        # ko puts it somewhere else... command: ['/usr/local/bin/cloud-controller-manager']
+        args: [] # args must be replaced by tooling
+        env:
+        - name: KUBERNETES_SERVICE_HOST
```
don't you need to set the ports too?
